### PR TITLE
Build with CUDA 12

### PR DESCRIPTION
**Pull request by pentschev**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Update UCX to build CUDA 12 from new packaging format and CUDA 10/11 with the legacy cudatoolkit package. Building with qemu segfaults, so switching back to Travis was necessary but that caused ppc64le builds to timeout and were thus disabled, at least until cross-compiling is stable (see #88).

Fixes https://github.com/conda-forge/nccl-feedstock/issues/82
Closes https://github.com/conda-forge/nccl-feedstock/pull/86
Closes https://github.com/conda-forge/nccl-feedstock/pull/89

### CODE REVIEW

**Comment by jakirkham** [comment]
@conda-forge-admin, please re-render

**Comment by leofang** [comment]
Seems to be compiling (on linux64)!

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,30 +1,40 @@
...
 requirements:
   build:
     - {{ compiler("c") }}
     - {{ compiler("cxx") }}
-    - {{ compiler("cuda") }}
+    - {{ compiler("cuda") }}         # [cuda_compiler_version != "12.0"]
+    - cuda-cudart ={{ cuda_compiler_version }}         # [cuda_compiler_version == "12.0"]
+    - cuda-cudart-dev ={{ cuda_compiler_version }}     # [cuda_compiler_version == "12.0"]
+    - cuda-cudart-static ={{ cuda_compiler_version }}  # [cuda_compiler_version == "12.0"]
+    - cuda-nvcc ={{ cuda_compiler_version }}           # [cuda_compiler_version == "12.0"]
```
Comment (h-vetinari) [inline recipe/meta.yaml:29]: Should these not all be run-deps of `cuda-nvcc` resp. `cuda-nvcc_<target>`? They look like compiler-support libs, I guess it's not the intention that every recipe would have to specify them by hand?

Same for the host-reqs (that are presumably there for the run-exports, which could reasonably be replaced by "strong" run-exports of the cuda-compiler). Compare e.g. the [strong run-exports](https://github.com/conda-forge/nvcc-feedstock/blob/main/recipe/meta.yaml#L40-L47) of the current nvcc metapackage.

PS. It's possible that this PR is only experimenting, but given that a migrator [PR](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/4400) now exists, I'm interested in how "real" CUDA 12 PRs would look like currently.

Comment (jakirkham) [reply recipe/meta.yaml:29]: Yeah they should (or really they should be dependencies of `cuda-nvcc_{{ target_platform }}`). We've been trying to dogfood the packages to learn about issues and fix them. This is certainly one such issue

The `requirements/host` dependencies here shouldn't be needed either

Yep, as you guessed, this is an experiment, but with the hope of getting something a bit more final (after some iteration on the aforementioned issues)

**Comment by leofang** [comment]
@conda-forge-admin, please rerender

**Comment by leofang** [comment]
Took the slimed-down migrator from https://github.com/conda-forge/cupy-feedstock/pull/199, and it rerendered to the same CI files, which is expected. Will test it in the UCX feedstock, before pushing to Bradley's migrator PR.

**Review by jakirkham — COMMENTED** [review]
Thanks Peter! 🙏 

Made a few minor suggestions for simplifications (particularly after fixes to the compiler's search paths)

File: conda-forge.yml [hunk 2]
```diff
@@ -2,14 +2,14 @@ conda_forge_output_validation: true
 os_version:
   linux_64: cos6
   linux_aarch64: cos7
-  linux_ppc64le: cos7
 provider:
-  linux_ppc64le: azure
-  linux_aarch64: azure
+  linux_aarch64: default
```
[the remark this answers is not in the document]
Comment (jakirkham) [reply conda-forge.yml:7]: (done below)

File: conda-forge.yml [hunk 3]
```diff
@@ -4,12 +4,14 @@ os_version:
   linux_aarch64: cos7
   linux_ppc64le: cos7
 provider:
-  linux_ppc64le: azure
-  linux_aarch64: azure
+  linux_aarch64: default
+  linux_ppc64le: default
```
Comment (jakirkham) [inline conda-forge.yml:8]: Sorry meant to keep the original settings (before this PR)

```suggestion
  linux_aarch64: azure
  linux_ppc64le: azure
```

**Comment by jakirkham** [comment]
@conda-forge-admin, please re-render

File: conda-forge.yml (hunk 3 above)
Comment (pentschev) [reply conda-forge.yml]: Ah ok, got it. Thanks John!

File: conda-forge.yml [hunk 4]
```diff
@@ -4,12 +4,14 @@ os_version:
   linux_aarch64: cos7
   linux_ppc64le: cos7
 provider:
-  linux_ppc64le: azure
   linux_aarch64: azure
+  linux_ppc64le: azure
```
Comment (jakirkham) [inline conda-forge.yml:8]: Preserving original order

```suggestion
  linux_ppc64le: azure
  linux_aarch64: azure
```

**Comment by jakirkham** [comment]
@conda-forge-admin, please re-render

File: conda-forge.yml (hunk 4 above)
Comment (pentschev) [reply conda-forge.yml]: Just curious, does this make a difference in practice or was it just to avoid unnecessary changes in the diff?

**Review by jakirkham — COMMENTED** [review]
Thanks Peter! 🙏

Had a few minor suggestions below

File: recipe/build.sh [hunk 5]
```diff
@@ -1,16 +1,24 @@
...
-    export NVCC_GENCODE="-gencode=arch=compute_60,code=[compute_60,sm_60] \
-                         -gencode=arch=compute_70,code=[compute_70,sm_70] \
-                         -gencode=arch=compute_80,code=[compute_80,sm_80]"
-    make -j${CPU_COUNT} src.lib CUDA_HOME="${CUDA_HOME}" CUDARTLIB="cudart_static" NVCC_GENCODE="$NVCC_GENCODE"
-else
-    make -j${CPU_COUNT} src.lib CUDA_HOME="${CUDA_HOME}" CUDARTLIB="cudart_static"
+    NVCC_GENCODE="-gencode=arch=compute_60,code=[compute_60,sm_60] \
+                  -gencode=arch=compute_70,code=[compute_70,sm_70] \
+                  -gencode=arch=compute_80,code=[compute_80,sm_80]"
+    EXTRA_ARGS="${EXTRA_ARGS} NVCC_GENCODE=\"${NVCC_GENCODE}\""
```
[the remark this answers is not in the document]
Comment (pentschev) [reply recipe/build.sh]: Done in https://github.com/conda-forge/nccl-feedstock/pull/85/commits/fbe7113177f1846cb115ae3564d154d2208d4416

**Comment by jakirkham** [comment]
@conda-forge-admin, please re-render

File: conda-forge.yml [hunk 6]
```diff
@@ -13,3 +13,5 @@ bot:
   automerge: true
 conda_build:
   pkg_format: '2'
+azure:
+  store_build_artifacts: True
```
Comment (jakirkham) [inline conda-forge.yml:17]: Think we can drop this as well

```suggestion
```

**Review by leofang — APPROVED** [review]

**Comment by jakirkham** [comment]
@conda-forge-admin, please re-render

File: recipe/meta.yaml [hunk 7]
```diff
@@ -1,17 +1,20 @@
...
-  sha256: f4ac3c74d469c9cd718f82e1477759785db9b9f8cc9d9ecc103485805b8394a3
+  sha256: 1311a6fd7cd44ad6d4523ba03065ce694605843fd30a5c0f77aa3d911abe706d
+  patches:
+    # Upstreaming w/PR: https://github.com/NVIDIA/nccl/pull/854
+    - 0001-Allow-custom-NVCC-path.patch
 
 build:
-  number: 1
-  skip: true  # [(not linux) or cuda_compiler_version in (undefined, "None")]
+  number: 2
```
Comment (jakirkham) [inline recipe/meta.yaml:16]: Since we are now bumping the version, this can be reset to `0`

```suggestion
  number: 0
```

File: recipe/build.sh [hunk 8]
```diff
@@ -1,7 +1,13 @@
 #!/bin/bash
 
+EXTRA_ARGS="CUDARTLIB=\"cudart_static\""
```
Comment (jakirkham) [inline recipe/build.sh:3]: Here's where `cudart_static` is specified

File: recipe/build.sh [hunk 9]
```diff
@@ -1,7 +1,13 @@
...
+
+if [[ "${cuda_compiler_version}" =~ 12.* ]]; then
+  EXTRA_ARGS="${EXTRA_ARGS} CUDA_HOME=\"${PREFIX}\" NVCC=\"${BUILD_PREFIX}/bin/nvcc\""
+elif [[ "${cuda_compiler_version}" != "None" ]]; then
+  EXTRA_ARGS="${EXTRA_ARGS} CUDA_HOME=\"${CUDA_PATH}\""
+fi
 
-EXTRA_ARGS="CUDA_HOME=\"${CUDA_HOME}\" CUDARTLIB=\"cudart_static\""
```
[the remark this answers is not in the document]
Comment (jakirkham) [reply recipe/build.sh:4]: Adding a review comment on that line for clarity on where it is set

**Review by jakirkham — COMMENTED** [review]
Just noting how `cudart_static` is set, nothing to do here :)

**Review by jakirkham — APPROVED** [review]

File: recipe/build.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (leofang) [reply recipe/build.sh:4]: Thanks, idk how I missed it.

**Review by leofang — APPROVED** [review]
Many thanks to you both, Peter & John!

**Pull request merged**